### ISSUE TITLE
E5113 "loop or previous error loading module" -- installing and setting colorscheme error

### ISSUE THREAD

**Issue by miltsghostrehab** [neovim/neovim#21865]
### Describe the bug

I just started converting my setup from Vimscript to Lua and found that if I use Lua to set `colorscheme` in the same script as I am to install it, I get an error. Once all plugins are installed, I don't get this error.

### Steps to reproduce

[nvim_loop_error.zip](https://github.com/neovim/neovim/files/10437865/nvim_loop_error.zip)

Minimal files to reproduce are included above. I can also reproduce it by using `colorscheme gruvbox` in the `vim.cmd` block.

Interestingly enough, this doesn't seem to be an issue if you copy/paste the contents of `pluggos.lua` into `init.lua`

### Expected behavior

I didn't expect the error, as this script worked fine in Vimscript.

### Neovim version (nvim -v)

NVIM v0.8.2

### Vim (not Nvim) behaves the same?

No

### Operating system/version

EndeavourOS

### Terminal name/version

alacritty 0.11.0 (8dbaa0bb)

### $TERM environment variable

alacritty

### Installation

Yay / Pacman

**Comment by clason** [comment]
Sorry, that is an issue with your configuration, not with Neovim: Why would you expect Neovim to be able to set a colorscheme (or, more generally, use a plugin) _before_ it is installed?

**Comment by miltsghostrehab** [comment]
> Sorry, that is an issue with your configuration, not with Neovim: Why would you expect Neovim to be able to set a colorscheme (or, more generally, use a plugin) _before_ it is installed?

As stated above:
1. This same configuration works fine when using Vimscript -- no errors thrown
2. If you copy/paste the contents of pluggos.vim into init.vim, everything works as expected -- no errors thrown.

Also, if you actually run these config files, it does what it should do -- installs the colorscheme then applies it. The error is unnecessary.

I am not expecting the colorscheme to be applied before it is installed. But if my configuration is wrong, then the error should be consistent.

**Comment by miltsghostrehab** [comment]
> I don't see why would it trigger that error, this shouldn't have anything to do with colorscheme? The only thing that could trigger it is `autocmd VimEnter * PlugInstall --sync | source $MYVIMRC`, or there is something more to your configuration that isn't included in the files.
> 
> Try changing `autocmd VimEnter * PlugInstall --sync | source $MYVIMRC` to
> 
> ```viml
> autocmd VimEnter * PlugInstall --sync | lua vim.schedule(function() vim.cmd('source $MYVIMRC') end)
> ```

I tried this new code, and the error remains. My personal init.lua has more to it, but I have pared it down to the smallest possible file to recreate this issue. I have no idea why it's only triggered with the colorscheme.

**Comment by ii14** [comment]
What's the full error message?

**Comment by miltsghostrehab** [comment]
```
Error executing vim.schedule lua callback: [string ":lua"]:1: Vim(source):E5113: Error while calling lua chunk: /home/lacy/.do
tfiles/nvim/init.lua:4: loop or previous error loading module 'pluggos'
stack traceback:
        [C]: in function 'require'
        /home/lacy/.dotfiles/nvim/init.lua:4: in main chunk
        [C]: in function 'cmd'
        [string ":lua"]:1: in function <[string ":lua"]:1>
stack traceback:
        [C]: in function 'cmd'
        [string ":lua"]:1: in function <[string ":lua"]:1>
```

**Comment by ii14** [comment]
This is with `vim.schedule`?

**Comment by ii14** [comment]
By the way, this `source $MYVIMRC` wouldn't work anyway. Lua modules are loaded only once.

**Comment by miltsghostrehab** [comment]
> This is with `vim.schedule`?

Yes.


> By the way, this `source $MYVIMRC` wouldn't work anyway. Lua modules are loaded only once.

This seems accurate, but I believe that `source $MYVIMRC` makes it so `vim-plug` is loaded, allowing for the subsequent plug-ins to be downloaded, installed and activated. The behavior seems different from what you're describing.


> I think this happens because `:colorscheme` fails loading the module, which allows it to be evaluated again, which in turn can make it require itself by re-running `init.lua`. If the module doesn't fail there is no problem.

I'm not sure if I'm getting it. Could you explain with a little more detail?

**Comment by ii14** [comment]
> I believe that `source $MYVIMRC` makes it so vim-plug is loaded, allowing for the subsequent plug-ins to be downloaded, installed and activated.

Nope. Let's ignore the error for now. If you don't have any errors in your lua module with plugins, it will run only once. It can run twice and activate your plugins only if the errors inside it happened to abort the script before. Keeping in mind how `require` works, making it work without errors is self contradictory and basically cannot work (I mean you could make it work, but then it would be pointless to use lua modules in the first place).

What you can try to do instead is get rid of `VimEnter` autocommand, and simply run `PlugInstall` inside that if statement, after you downloaded vim-plug. It should block your configuration from proceeding, and once it ends everything should be available.

**Comment by miltsghostrehab** [comment]
> Nope. Let's ignore what the error is for now. If you don't have any errors in your lua module with plugins, it will run only once. It can run twice and activate your plugins only if the errors inside it happened to abort the script before. Keeping in mind how `require` works, making it work without errors is self contradictory and basically cannot work (I mean you could make it work, but then it would be pointless to use lua modules in the first place).
> 
> What you can try to do instead is get rid of `VimEnter` autocommand, and simply run `PlugInstall` inside that if statement, after you downloaded vim-plug. It should block your configuration from proceeding, and once it ends everything should be available.

Ok, I think I get it! Lua allows for a module to be run a second time if errors are thrown which abort the script. Would that explain why the error is thrown even though the colorscheme is applied?

**Comment by ii14** [comment]
I'm still not sure what's the exact sequence here, but that's probably it? Throws an error about unknown color scheme initially, which also triggers the "loop or previous error" error, and then the second time it works and the color scheme is applied?
